File: typescript/src/main/vsdk/toolkit/io/image/PicOutputHandle.ts
import { File } from "../../../../java/io/File";
import { FileOutputStream } from "../../../../java/io/FileOutputStream";
import { OutputStream } from "../../../../java/io/OutputStream";
import { ColorRgb } from "../../common/ColorRgb";
import { DkColor } from "./DkColor";
import { ImageOutputHandle } from "./ImageOutputHandle";

const util = require("node:util");

export class PicOutputHandle extends ImageOutputHandle {
  private outputStream: OutputStream | null;

  private static formatToString(format: string | null, ...argumentsList: unknown[]): string {
    if (format === null) {
      return "";
    }
    try {
      return util.format(format, ...argumentsList);
    }
    catch (_ignored) {
      return "";
    }
  }

  private static writeFormatted(
    outputStream: OutputStream | null,
    format: string | null,
    ...argumentsList: unknown[]
  ): void {
    if (outputStream === null || format === null) {
      return;
    }

    const text = PicOutputHandle.formatToString(format, ...argumentsList);
    if (text.length <= 0) {
      return;
    }

    const bytes = new Uint8Array(Buffer.from(text, "ascii"));
    try {
      outputStream.write(bytes, 0, bytes.length);
    }
    catch (_ignored) {
    }
  }

  private static formatCompileDate(date: Date): string {
    const monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
    const month = monthNames[date.getMonth()];
    const day = `${date.getDate()}`.padStart(2, "0");
    return `${month} ${day} ${date.getFullYear()}`;
  }

  public constructor(filename: string, w: number, h: number) {
    super();
    this.init("high dynamic range PIC", w, h);
    this.outputStream = null;

    const file = new File(filename);
    const canWrite = file.canWrite();
    const isDirectory = file.isDirectory();
    file.dispose();

    if (!canWrite || isDirectory) {
      this.outputStream = null;
      process.stderr.write("Can't open PIC output");
      return;
    }

    try {
      this.outputStream = new FileOutputStream(filename);
    }
    catch (_ignored) {
      this.outputStream = null;
      process.stderr.write("Can't open PIC output");
      return;
    }

    this.writeHeader();
  }

  public closeHandle(): void {
    if (this.outputStream !== null) {
      try {
        this.outputStream.close();
      }
      catch (_ignored) {
      }
    }
    this.outputStream = null;
  }

  public override writeRadianceRGB(rgbRadiance: ColorRgb[]): number {
    let result = 0;
    if (this.outputStream !== null) {
      result = DkColor.writeScan(rgbRadiance, this.width, this.outputStream);
    }

    if (result !== 0) {
      return this.width;
    }
    return 0;
  }

  private writeHeader(): void {
    PicOutputHandle.writeFormatted(this.outputStream, "#?RADIANCE\n");
    const compileDate = PicOutputHandle.formatCompileDate(new Date());
    PicOutputHandle.writeFormatted(this.outputStream, "#RPK PicOutputHandler (compiled %s)\n", compileDate);
    PicOutputHandle.writeFormatted(this.outputStream, "FORMAT=32-bit_rle_rgbe\n");
    PicOutputHandle.writeFormatted(this.outputStream, "\n");
    PicOutputHandle.writeFormatted(this.outputStream, "-Y %d +X %d\n", this.height, this.width);
  }
}

File: turboc/base/src/io/image/PicOutputHandle.cpp
#include <stdarg.h>

#include "java/io/FileOutputStream.h"
#include "java/lang/System.h"
#include "java/util/Formatter.h"
#include "io/wrapper/PersistenceElement.h"
#include "io/image/Dkcolor.h"
#include "io/image/PicOutputHandle.h"

#ifndef va_copy
#if defined(__va_copy)
#define va_copy(dst, src) __va_copy((dst), (src))
#else
#define va_copy(dst, src) ((dst) = (src))
#endif
#endif

String
PicOutputHandle::formatToString(const char *format, va_list arguments) {
    if ( format == NULL ) {
        return String();
    }

    char localBuffer[256];
    va_list argumentsCopy;
    va_copy(argumentsCopy, arguments);
    const int required = Formatter::vformat(localBuffer, ((int)(sizeof(localBuffer))), format, argumentsCopy);
    va_end(argumentsCopy);

    if ( required < 0 ) {
        return String();
    }
    if ( required < ((int)(sizeof(localBuffer))) ) {
        return String(localBuffer);
    }

    char *dynamicBuffer = new char[required + 1];
    va_copy(argumentsCopy, arguments);
    Formatter::vformat(dynamicBuffer, required + 1, format, argumentsCopy);
    va_end(argumentsCopy);

    String result(dynamicBuffer);
    delete[] dynamicBuffer;
    return result;
}

void
PicOutputHandle::writeFormatted(OutputStream *outputStream, const char *format, ...) {
    if ( outputStream == NULL || format == NULL ) {
        return;
    }

    va_list arguments;
    va_start(arguments, format);
    String text = formatToString(format, arguments);
    va_end(arguments);

    if ( text.isEmpty() ) {
        return;
    }
    PersistenceElement::writeBytes(
        *outputStream,
        ((const unsigned char *)(text.toCString())),
        text.length());
}

PicOutputHandle::PicOutputHandle(const char *filename, int w, int h) {
    ImageOutputHandle::init("high dynamic range PIC", w, h);

    File file(filename);
    if ( !file.canWrite() || file.isDirectory() ) {
        outputStream = NULL;
        System::err.printf("Can't open PIC output");
        return;
    }
    outputStream = new FileOutputStream(filename);

    writeHeader();
}

PicOutputHandle::~PicOutputHandle() {
    if ( outputStream != NULL ) {
        outputStream->close();
        delete outputStream;
    }
    outputStream = NULL;
}

/**
Writes scanline of high-dynamic range radiance data in RGB format
*/
int
PicOutputHandle::writeRadianceRGB(const ColorRgb *rgbRadiance) {
    int result = 0;

    if ( outputStream != NULL ) {
        float *scanline = new float[width * 3];
        for ( int i = 0; i < width; i++ ) {
            scanline[3 * i] = rgbRadiance[i].getR();
            scanline[3 * i + 1] = rgbRadiance[i].getG();
            scanline[3 * i + 2] = rgbRadiance[i].getB();
        }
        result = DkColor::writeScan(((DK_COLOR *)(scanline)), width, outputStream);
        delete[] scanline;
    }

    if ( result ) {
        return width;
    } else {
        // We don't know how many pixels were actually written
        return 0;
    }
}

void
PicOutputHandle::writeHeader() {
    // Simple RADIANCE header
    writeFormatted(outputStream, "#?RADIANCE\n");
    writeFormatted(outputStream, "#RPK PicOutputHandler (compiled %s)\n", __DATE__);
    writeFormatted(outputStream, "FORMAT=32-bit_rle_rgbe\n");
    writeFormatted(outputStream, "\n");
    writeFormatted(outputStream, "-Y %d +X %d\n", height, width);
}
